csrc/sycl/ops.h

```diff
@@ -0,0 +1,215 @@
+				}
+
+};
+
+class ContextCusparse
+{
+    public:
+                                sycl::queue *m_handle;
+
+                                ContextCusparse()
+				{
+                                        sycl::queue *handle;
+                                        handle = &dpct::get_default_queue();
+                                        m_handle = handle;
+				}
+
+};
+
+
+template <typename T> void estimateQuantiles(T *A, float *code, float offset, int n);
+
+void quantize(float *code, float *A, unsigned char *out, int n);
+void dequantize(float *code, unsigned char *A, float *out, int n);
+template <typename T, int STOCHASTIC, int DATA_TYPE> void quantizeBlockwise(float * code, T *A, float *absmax, unsigned char *out, float* rand, int rand_offset, int blocksize, const int n);
+template<typename T, int DATA_TYPE> void dequantizeBlockwise(float *code, unsigned char *A, float *absmax, T *out, int block_size, const int n);
+
+template<typename T, int OPTIMIZER> void optimizer32bit(T* g, T* p,
+                float* state1, float* state2, float *unorm, float max_unorm, float param_norm,
+                float beta1, float beta2, float eps, float weight_decay,
+                int step, float lr, const float gnorm_scale, bool skip_zeros, int n);
+
+template<typename T, int OPTIMIZER> void optimizerStatic8bit(T* p, T* g, unsigned char* state1, unsigned char* state2,
+                float *unorm, float max_unorm, float param_norm,
+                float beta1, float beta2,
+                float eps, int step, float lr,
+                float* quantiles1, float* quantiles2,
+                float* max1, float* max2, float* new_max1, float* new_max2,
+                float weight_decay,
+                const float gnorm_scale, int n);
+
+template<typename T, int OPTIMIZER> void optimizerStatic8bitBlockwise(T* p, T* g,
+                unsigned char* state1, unsigned char* state2, float beta1, float beta2, float eps, int step, float lr,
+                float* quantiles1, float* quantiles2, float* absmax1, float* absmax2, float weight_decay, const float gnorm_scale,
+								bool skip_zeros, int n);
+
+template<typename T> void percentileClipping(T * g, float *gnorm_vec, int step, const int n);
+
+extern SYCL_EXTERNAL void histogramScatterAdd2D(float* histogram, int *index1, int *index2, float *src, int maxidx1, int n);
+
+void gemmex(Context * context, bool transposeA, bool transposeB, int m, int n, int k, void *A, void *B, void *C, int lda, int ldb, int ldc);
+void strided_gemmex(Context *context, bool transposeA, bool transposeB, int m, int n, int k, void *A, void *B, void *C, int lda, int ldb, int ldc,
+                    long long int strideA, long long int strideB, long long int strideC, int batchCount);
+
+
+template <int FORMATB, int DTYPE_OUT, int SCALE_ROWS> int igemmlt(dpct::queue_ptr ltHandle, int m, int n, int k, const int8_t *A, const int8_t *B, void *C, float *row_scale, int lda, int ldb, int ldc);

```

Comment:
Essentially the whole thing results on the key question of whether there are any existing libraries like oneDNN/oneMKL blas etc that support a matrix "order" enum that we can assign appropriate values to to coincide with cublasLtOrder_t, such that e.g. oneDNN cuda backend can then call the appropriate cublaslt interface with the appropriate native enum value.

Without this it will not be possible to call the appropriate cublaslt function. If you are lucky and such an interface already exists in oneDNN, then eventually there will be applications that require more of the logical structure exposed in https://docs.nvidia.com/cuda/cublas/#using-the-cublaslt-api

Essentially you need a library that supports things like matrix format arch generation specific enums for arch generation specific specialization of kernels etc.

Without this we won't be able to do much with this. Until this question has been addressed I don't really see much point in us reviewing this further.

Reply:
@JackAKirk For a SYCL+CUDA target, wouldn't we be able to just specialize here and use cuBLASLt directly?
For other targets, does this matter? We could restrict to the equivalent of `CUBLASLT_ORDER_COL` and `CUBLASLT_ORDER_ROW` if we had to, I would think?

Reply:
> @JackAKirk For a SYCL+CUDA target, wouldn't we be able to just specialize here and use cuBLASLt directly?
>
I'm not sure if you could currently use it directly even if you don't care about specializing for a particular nvidia gpu generation: this may be possible since bitsandbytes doesn't seem to use many of the other enums exposed in cublaslt that are e.g. used for problem specific implementation decisions: an example here is the epilogue type.
But even if you can call cublaslt assuming e.g. AMPERE (sm_80), then you won't be able to also make it fast on TURING (where the optimized impl is completely different), unless whatever sycl wrapper library you use exposes the arch specific enum values that cublaslt uses.

> For other targets, does this matter? We could restrict to the equivalent of `CUBLASLT_ORDER_COL` and `CUBLASLT_ORDER_ROW` if we had to, I would think?

For other targets you may not be able to reuse much of the code (I haven't thought about this deeply, and I may be wrong here, but I'd be surprised if you could reuse much) since this is all specialized to nvidia hardware (and indeed specific arch generations of nvidia hardware). The matrix multiplication hardware on intel gpus is completely different. They support completely different sizes of low level gemm operations.

Reply:
> @JackAKirk For a SYCL+CUDA target, wouldn't we be able to just specialize here and use cuBLASLt directly? For other targets, does this matter? We could restrict to the equivalent of `CUBLASLT_ORDER_COL` and `CUBLASLT_ORDER_ROW` if we had to, I would think?

I think you have to compare the cublaslt specification (https://docs.nvidia.com/cuda/cublas/#using-the-cublaslt-api) with the oneDNN spec carefully. This is is the only way to answer such questions. The point I made is that until the answers to these questions are known, I don't see how much progress can be made.

Reply:
On the python side, bitsandbytes already does specialize:

```python
def get_special_format_str():
    if not torch.cuda.is_available():
        return "col_turing"
    major, _minor = torch.cuda.get_device_capability()
    if major <= 7:
        return "col_turing"
    if major == 8:
        return "col_ampere"
    return "col_turing"
```
The `bitsandbytes.functional.transform()` function is used (which calls the C++ library `ctransform_row2turing, ctransform_row2ampere, etc). Then we have specific implementations in the C++ library: `cigemmlt_turing_32, cigemmlt_turing_8, cigemmlt_ampere_32, cigemmlt_ampere_8`

Reply:
> @JackAKirk For a SYCL+CUDA target, wouldn't we be able to just specialize here and use cuBLASLt directly? For other targets, does this matter? We could restrict to the equivalent of `CUBLASLT_ORDER_COL` and `CUBLASLT_ORDER_ROW` if we had to, I would think?

Oh sorry I think I misunderstood what you meant at first: previously it had been stated to me that cublaslt calls would be ported to some sycl library irrespective of the target.
I think that the option of calling native cuda libraries in the sycl (cuda) backend would also be an option that could be investigated. I think this is what you mean.

Reply:
I do tend to agree though, outside of NVIDIA hardware there will need to be new implementations for this that I wouldn't expect e.g. SYCLomatic / DPC++ migration tool to be able to handle. I'm not really sure what it looks like for HIP either.

There's lots to do, so one step at a time.

Reply:
That's essentially what I mean.

Point of further clarification: I don't expect we would really depart from the native CUDA implementation we have today as a default for NVIDIA hardware. The primary use case for SYCL should be around non-NVIDIA hardware. That said I don't see why we couldn't have a mix of SYCL+CUDA code as an option, maybe experimentally, but IMHO it shouldn't be a priority.

Reply:
> That's essentially what I mean.
> 
> Point of further clarification: I don't expect we would really depart from the native CUDA implementation we have today as a default for NVIDIA hardware. The primary use case for SYCL should be around non-NVIDIA hardware. 

Totally makes sense to me. I'm just reviewing in the capacity my team has been asked to, which is regards to enabling a port that supports the cuda backend and allows good performance.

Reply:
Yes on this topic, there are a couple of approaches (since gemms are not ported yet). onemkl blas porting mirroring the cublaslt apis , which would involve segregation of descriptors as it would not be a 1:1 mapping. An alternate route is oneDNN . Now since cublasLT is closed source, it is difficult to ascertain the improvements met from using cublasLt over cudnn calls for gemm.  In terms of SYCL support of cublasLt, it is a WIP to integrate similar interface . For ptx backends, I think that the SYCL api (when made available) for cublas would be compatible with nv devices, whereas an option for keeping cublas calls as it is when building sycl code sounds good as well. 
But since primarily we plan to integrate intel sycl support, I think that a translation is must needed in this case.